blob/blob.go

```diff
@@ -18,42 +21,107 @@ import (
-// TODO (@vgonkivs): rework `Proof` in order to prove a particular blob.
-// https://github.com/celestiaorg/celestia-node/issues/2303
-type Proof []*nmt.Proof
-
-func (p Proof) Len() int { return len(p) }
+// Proof constructs the proof of a blob to the data root.
+type Proof struct {
+	// ShareToRowRootProof the proofs of the shares to the row roots they belong to.
+	// If the blob spans across multiple rows, then this will contain multiple proofs.
+	ShareToRowRootProof []*tmproto.NMTProof
+	// RowProof the proofs of the row roots containing the blob shares
+	// to the data root.
+	RowProof coretypes.RowProof
+}
 
 // equal is a temporary method that compares two proofs.
 // should be removed in BlobService V1.
-func (p Proof) equal(input Proof) error {
-	if p.Len() != input.Len() {
-		return ErrInvalidProof
-	}
-
-	for i, proof := range p {
-		pNodes := proof.Nodes()
-		inputNodes := input[i].Nodes()
-		for i, node := range pNodes {
-			if !bytes.Equal(node, inputNodes[i]) {
+func (p *Proof) equal(input *Proof) error {
+	// compare the NMT proofs
+	for i, proof := range p.ShareToRowRootProof {
+		if proof.Start != input.ShareToRowRootProof[i].Start {
+			// TODO(@rach-id): should we define specific errors for each case? that's better
+			// to know which part is not equal.
+			// Also, this error should be ErrProofNotEqual or similar, and not ErrInvalidProof
+			return ErrInvalidProof
+		}
+		if proof.End != input.ShareToRowRootProof[i].End {
+			// TODO(@rach-id): should we define specific errors for each case? that's better
+			// to know which part is not equal.
+			// Also, this error should be ErrProofNotEqual or similar, and not ErrInvalidProof
+			return ErrInvalidProof
+		}
+		for j, node := range proof.Nodes {
+			if !bytes.Equal(node, input.ShareToRowRootProof[i].Nodes[j]) {
 				return ErrInvalidProof
 			}
 		}
+	}
 
-		if proof.Start() != input[i].Start() || proof.End() != input[i].End() {
+	// compare the row proof
+	if p.RowProof.StartRow != input.RowProof.StartRow {
+		return ErrInvalidProof
+	}
+	if p.RowProof.EndRow != input.RowProof.EndRow {
+		return ErrInvalidProof
+	}
+	for index, rowRoot := range p.RowProof.RowRoots {
+		if !bytes.Equal(rowRoot, input.RowProof.RowRoots[index]) {
 			return ErrInvalidProof
 		}
-
-		if !bytes.Equal(proof.LeafHash(), input[i].LeafHash()) {
+	}
+	for i, proof := range p.RowProof.Proofs {
+		if proof.Index != input.RowProof.Proofs[i].Index {
+			return ErrInvalidProof
+		}
+		if proof.Total != input.RowProof.Proofs[i].Total {
 			return ErrInvalidProof
 		}
+		for j, aunt := range proof.Aunts {
+			if !bytes.Equal(aunt, input.RowProof.Proofs[i].Aunts[j]) {
+				return ErrInvalidProof
+			}
+		}
 	}
 	return nil
 }
 
+// Verify takes a set of shares, a namespace and a data root, and verifies if the
+// provided shares are committed to by the data root.
+func (p *Proof) Verify(rawShares [][]byte, namespace share.Namespace, dataRoot []byte) (bool, error) {
```

Comment:
We will also add another method that will verify this proof but using the subtree roots

Reply:
https://github.com/celestiaorg/celestia-node/issues/3611